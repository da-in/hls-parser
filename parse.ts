import {Playlist, TAG} from './types';

/**
 * @param text m3u8 string
 */
const parse = (text: string): Partial<Playlist> | undefined => {
    const playlist = {}
    const tags: TAG[] = []
    const lines = text.split('\n').map((line) => line.trim())
    let type

    // not a extended m3u
    if(lines[0] !== '#EXTM3U'){
        return
    }

    for (let i = 1; i < lines.length; i++){
        const line = lines[i]
        const re: RegExp = /#([A-Z0-9-]+):?([\w\d=,-]+)?/
        const match = line.match(re)

        if(!match){
            continue
        }

        const name = match[1]
        const attributes = parseAttributes(match[2])

        switch(name){
            // master playlist: 동일한 컨텐츠에 대해 다른 인코딩 버전을 제공
            case 'EXT-X-MEDIA': // 다른 유형의 미디어(오디오, 비디오, 자막)와 추가 정보
            case 'EXT-X-STREAM-INF': // 다른 변형 스트림(Variant Stream)
            case 'EXT-X-I-FRAME-STREAM-INF':
            case 'EXT-X-SESSION-DATA':
            case 'EXT-X-SESSION-KEY':
                type = 'MasterPlaylist'
        }

        tags.push({
            name,
            attributes,
        })
    }

    // master playlist tags

    // #EXT-X-MEDIA
    // #EXT-X-STREAM-INF
    // #EXT-X-I-FRAME-STREAM-INF
    // #EXT-X-SESSION-DATA
    // #EXT-X-SESSION-KEY

    Object.assign(playlist, {type, tags})
    return playlist
}

/**
 * @param attributesText attributes string separated with ','
 */
const parseAttributes = (attributesText: string): TAG['attributes'] => {
    const result: Record<string, string> = {}
    attributesText.split(',').map((attribute) => {
        const [key, value] = attribute.split('=')
        result[key] = value
    })
    return result
}

export default parse
