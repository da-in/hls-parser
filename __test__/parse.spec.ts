import * as fs from 'fs'
import path from 'node:path'
import {describe, it, expect} from '@jest/globals'
import parse from '../parse'

const loadFile = (file: string) => {
    const _path = path.join(__dirname, './', file)
    return fs.readFileSync(_path, 'utf8')
}

describe('parse', () => {
    it('parse master playlist', () => {
        const m3u8 = loadFile('./skate_phantom_flex_4k.m3u8')
        const playlist = parse(m3u8)
        expect(playlist?.type).toBe('MasterPlaylist')
        expect(playlist).toMatchSnapshot()
    })
    it('parse media playlist', () => {
    })
})
