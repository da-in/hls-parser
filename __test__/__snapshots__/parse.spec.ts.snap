// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse parse master playlist 1`] = `
{
  "tags": [
    {
      "attributes": {
        "BANDWIDTH": "288000",
        "PROGRAM-ID": "1",
        "RESOLUTION": "256x144",
      },
      "name": "EXT-X-STREAM-INF",
    },
    {
      "attributes": {
        "BANDWIDTH": "582000",
        "PROGRAM-ID": "1",
        "RESOLUTION": "426x240",
      },
      "name": "EXT-X-STREAM-INF",
    },
    {
      "attributes": {
        "BANDWIDTH": "1018000",
        "PROGRAM-ID": "1",
        "RESOLUTION": "640x360",
      },
      "name": "EXT-X-STREAM-INF",
    },
    {
      "attributes": {
        "BANDWIDTH": "1366000",
        "PROGRAM-ID": "1",
        "RESOLUTION": "854x480",
      },
      "name": "EXT-X-STREAM-INF",
    },
    {
      "attributes": {
        "BANDWIDTH": "2742000",
        "CODECS": "",
        "PROGRAM-ID": "1",
      },
      "name": "EXT-X-STREAM-INF",
    },
    {
      "attributes": {
        "BANDWIDTH": "5400000",
        "CODECS": "",
        "PROGRAM-ID": "1",
      },
      "name": "EXT-X-STREAM-INF",
    },
    {
      "attributes": {
        "BANDWIDTH": "7556000",
        "CODECS": "",
        "PROGRAM-ID": "1",
      },
      "name": "EXT-X-STREAM-INF",
    },
    {
      "attributes": {
        "BANDWIDTH": "10768000",
        "CODECS": "",
        "PROGRAM-ID": "1",
      },
      "name": "EXT-X-STREAM-INF",
    },
  ],
  "type": "MasterPlaylist",
}
`;
